internal/auth/auth.go

```diff
@@ -54,3 +41,47 @@ func RequiresTokenCheck(ctx context.Context, _ any, next graphql.Resolver) (any,
 	}
 	return next(ctx)
 }
+
+type PrivilegeContractValidator struct {
+	VehicleNFTAddress common.Address
+	// ManufAddr         string
+}
+
+// VehicleNFTPrivCheck checks if the claim set in the context includes the correct address the required privileges for the VehicleNFT contract.
+func (pcv *PrivilegeContractValidator) VehicleNFTPrivCheck(ctx context.Context, _ any, next graphql.Resolver, privs []model.Privilege) (any, error) {
+	claim, err := getTelemetryClaim(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", errUnauthorized.Error(), err)
+	}
+
+	if claim.ContractAddress != pcv.VehicleNFTAddress {
+		return nil, fmt.Errorf("%w: contract address mismatch", errUnauthorized)
+	}
+
+	for _, priv := range privs {
+		if _, ok := claim.privileges[priv]; !ok {
+			return nil, fmt.Errorf("%w: missing required privilege %s", errUnauthorized, priv)
+		}
+	}
+
+	return next(ctx)
+}
+
+// func (pcv *PrivilegeContractValidator) ManufacturerNFTPrivCheck(ctx context.Context, _ any, next graphql.Resolver, privs []model.Privilege) (any, error) {
```

Comment:
When we implement this I would like to combine the duplicate logic in its own function.